In [36]:
from sqlalchemy import create_engine, MetaData, Table
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, String, Date, ForeignKey

metadata = MetaData()

db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"
db = create_engine(db_string)

db_string1 = "mysql://bauer1:W0JjW9jge2SYwwYh@mysql.agh.edu.pl:3306/bauer1"
db1 = create_engine(db_string1)

In [51]:
session = (sessionmaker(bind=db))()
session1 = (sessionmaker(bind=db1))()
Base = declarative_base()

### <span style='color:Yellow'> 1. How many categories of films we have in the rental? </span>

In [38]:
class Category(Base):
    __tablename__ = 'category'
    category_id  = Column(Integer, primary_key=True)
    name = Column(String(50))
    last_update  = Column(Date)
    def __str__(self):
        return 'Category id:{0}\nCategory name: {1}\nCategory last_update: {2}'.format(self.category_id,self.name,self.last_update)

In [39]:
session_stmt = session.query(Category.category_id).count()
print(session_stmt)

16


### <span style='color:red'> Answer: </span> 16 categories

### <span style='color:Yellow'> 2. Display list of categories in alphabetic order. </span>

In [40]:
session_stmt = session.query(Category.name).order_by(Category.name)
session_results = session_stmt.all()
print(session_stmt)
session_results

SELECT category.name AS category_name 
FROM category ORDER BY category.name


[('Action',),
 ('Animation',),
 ('Children',),
 ('Classics',),
 ('Comedy',),
 ('Documentary',),
 ('Drama',),
 ('Family',),
 ('Foreign',),
 ('Games',),
 ('Horror',),
 ('Music',),
 ('New',),
 ('Sci-Fi',),
 ('Sports',),
 ('Travel',)]

### <span style='color:Yellow'> 3. Find the oldest and youngest film in rental. </span>

In [41]:
class Film(Base):
    __tablename__ = 'film'
    film_id  = Column(Integer, primary_key=True)
    title = Column(String(50))
    release_year  = Column(Date)
    def __str__(self):
        return 'Film id:{0}\nFilm title: {1}\nFilm release_year: {2}'.format(self.film_id,self.title,self.release_year)

In [44]:
from sqlalchemy import func 
session_stmt = session.query(func.max(Film.release_year)) 
print(session_stmt)
session_results = session_stmt.all()
print(session_results)

SELECT max(film.release_year) AS max_1 
FROM film
[(2006,)]


In [45]:
session_stmt = session.query(func.min(Film.release_year)) 
print(session_stmt)
session_results = session_stmt.all()
print(session_results)

SELECT min(film.release_year) AS min_1 
FROM film
[(2006,)]


### <span style='color:red'> Answer: </span> Both earliest and oldest film are year 2006.

### <span style='color:Yellow'> 4. How many rentals were in between 2005-07-01 and 2005-08-01? </span>

In [46]:
class Rental(Base):
    __tablename__ = 'rental'
    rental_id  = Column(Integer, primary_key=True)
    rental_date  = Column(Date)
    def __str__(self):
        return 'Rental id:{0}\nRental date: {1}'.format(self.rental_id,self.rental_date)

In [47]:
from sqlalchemy import or_, and_

session_stmt = session.query(Rental.rental_date).filter(and_(Rental.rental_date > '2005-07-01', Rental.rental_date < '2005-08-01')).count()
print(session_stmt)

6709


### <span style='color:red'> Answer: </span> 6709

### <span style='color:Yellow'> 5. How many rentals were in between 2010-01-01 and 2011-02-01? </span>

In [48]:
from sqlalchemy import or_, and_

session_stmt = session.query(Rental.rental_date).filter(and_(Rental.rental_date > '2010-01-01', Rental.rental_date < '2011-02-01')).count()
print(session_stmt)

0


### <span style='color:red'> Answer: </span> 0

### <span style='color:Yellow'> 6. Find the biggest payment in the rental. </span>

In [52]:
class Payments(Base):
    __tablename__ = 'payments'
    customerNumber  = Column(Integer, primary_key=True)
    amount   = Column(Integer)
    def __str__(self):
        return 'Payments customerNumber:{0}\nPayments amount: {1}'.format(self.customerNumber,self.amount )

In [53]:
session_stmt = session1.query(func.max(Payments.amount)) 
print(session_stmt)
session_results = session_stmt.all()
print(session_results)

SELECT max(payments.amount) AS max_1 
FROM payments
[(Decimal('120166.58'),)]


### <span style='color:red'> Answer: </span> $120166.58

### <span style='color:Yellow'> 7. Find all customers from Poland or Nigeria or Bangladesh. </span>

In [58]:
class Customers(Base):
    __tablename__ = 'customers'
    customerNumber  = Column(Integer, primary_key=True)
    customerName = Column(String(50))
    country   = Column(String(50))
    def __str__(self):
        return 'Customer number:{0}\nCustomer name: {1}\nCountry: {2}'.format(self.customerNumber,self.customerName,self.country)

In [61]:
session_stmt = session1.query(Customers.country).filter(or_(Customers.country == 'Poland', Customers.country == 'Nigeria', Customers.country == 'Bangladesh'))
print(session_stmt)
session_results = session_stmt.all()
print(session_results)

SELECT customers.country AS customers_country 
FROM customers 
WHERE customers.country = %s OR customers.country = %s OR customers.country = %s
[('Poland',)]
